solarforecastarbiter/io/reference_observations/arm.py

```diff
@@ -176,3 +231,151 @@ def update_observation_data(api, sites, observations, start, end):
             api, partial(fetch, doe_arm_user_id=doe_arm_user_id,
                          doe_arm_api_key=doe_arm_api_key),
             site, observations, start, end)
+
+
+def adjust_site_parameters(site):
+    """Updates extra parameters with applicable datastreams from
+    `arm_reference_sites.json`
+
+    Parameters
+    ----------
+    site: dict
+
+    Returns
+    -------
+    dict
+        Copy of input with updated extra parameters.
+    """
+    with open(DEFAULT_SITEFILE) as fp:
+        sites_metadata = json.load(fp)['sites']
+
+    # ARM has multiple 'locations' at each 'site'. In the Solar Forecast
+    # Arbiter we store each 'location' as a site. We use the `network_api_id`
+    # to indicate location, and `network_api_abbreviation` together to indicate
+    # the arm site. This is necessary to use both because location ids are only
+    # unique for a given site.
+    arm_location_id = site['extra_parameters']['network_api_id']
+    arm_site_id = site['extra_parameters']['network_api_abbreviation']
+
+    for site_metadata in sites_metadata:
+        site_extra_params = json.loads(site_metadata['extra_parameters'])
+        if (
+            site_extra_params['network_api_id'] == arm_location_id
+            and site_extra_params['network_api_abbreviation'] == arm_site_id
+
+        ):
+            site_out = site.copy()
+            site_out['extra_parameters'] = site_extra_params
+            return site_out
+    return site
+
+
+def find_stream_data_availability(streams, start, end):
+    """Determines what date ranges to use for each datastream.
+
+    Parameters
+    ----------
+    streams: dict
+        Dict where values are string datastream names and values are iso8601
+        date ranges `start/end` indicating the period of data available at
+        that datastream.
+    start: datetime
+        The start of the period to request data for.
+    end: datetime
+        The end of the period to request data for.
+
+    Returns
+    -------
+    dict:
+        Dict where keys are datastreams and values are two element lists of
+        `[start datetime, end datetime]` that when considered together should
+        span all of the available data between the requested start and end.
+    """
+    stream_range_dict = {}
+
+    # Find the overlap between each streams available data, and the requested
+    # period
+    for datastream, date_range in streams.items():
+        stream_range = parse_iso_date_range(date_range)
+        overlap = get_period_overlap(
+            start, end, stream_range[0], stream_range[1])
+        if overlap is None:
+            # The datastream did not contain any data within the requested
+            # range, we don't need to use it for this request.
+            continue
+        else:
+            stream_range_dict[datastream] = overlap
+
+    # Remove any overlap between streams
+    prev_start = None
+    prev_end = None
+    for datastream, date_range in stream_range_dict.items():
```

Comment:
this is confusing to me. As I understand it, the goal is to restrict each datastream from start to end, but also to prefer newer datastreams when they overlap. is that right? Or does the ordering in the streams dict matter? What if they overlap completely?

Reply:
see the new test. I also don't like modifying the list inside the dict in place. I would think the first for loop in this function would return a list of (datastream, [start, end]) and then you can iterate and assign to stream_range_dict rather than modifying

Reply:
> this is confusing to me. As I understand it, the goal is to restrict each datastream from start to end, but also to prefer newer datastreams when they overlap. is that right? Or does the ordering in the streams dict matter? What if they overlap completely?

The ordering in the stream dicts matter as it's currently written, with no preference for newer streams. Complete overlap is not something I had considered here, good catch. It would probably result in one stream being set with start/end both set to start.

My thoughts were that this would be a good way to handle more datastreams should we need them. But I'm seeing now that this would be fragile and prone to breaking if the date ranges weren't configured in order.

Perhaps the way to do this is just manually configure the datastreams to when we want to use them rather than trying to programatically select during the request based on what's available. We'd still need to determine overlap with the requested period, but that's a much simpler problem. What do you think?

Either way, I agree this would be cleaner without mangling the dict in place.

Reply:
Yes, I agree that just choosing not to have any overlap when adding to the json file is the way to go. Much easier than dynamically determining. maybe add some comment field with that info to the file